### PR TITLE
Wildcarded users

### PR DESCRIPTION
**Pull request by ilejn**
## Description

Implements Wildcarded users concept https://hackmd.io/LXk8tLI8Q3etKax-HWtGxQ
as agreed in https://github.com/ContentSquare/chproxy/pull/170

The idea is to have chproxy users with a suffix, that is out if chproxy control,
like analyst_* . 
For such users password from request to chproxy is retransmitted to ClickHouse as is.
So, analyst_jane and analyst_bob can send their requests to ClickHouse.

Rationale:  avoid nececity to have all users listed in chproxy config, ability to use LDAP or kerberos facilities at ClickHouse side.

## Pull request type

Please check the type of change your PR introduces:
- [x ] Feature

### Checklist

- [ x] Linter passes correctly
- [ ] Add tests which fail without the change (if possible)
- [x ] All tests passing
- [ x] Extended the README / documentation, if necessary

golangci-lint 1.49.0 reports several errors that are not related to the change
## Does this introduce a breaking change?

 [x ] No

## Further comments

I need some advice on how to add a test that actually covers the new functionality

### CODE REVIEW

**Comment by mga-chka** [comment]
Hi,
Glad to see this PR!

I haven't look at your PR yet but regarding your question you can add some tests in the proxy_test.go file.
This file already contains tests about authentication, e.g: https://github.com/ContentSquare/chproxy/blob/master/proxy_test.go#L310
You could add 2-3 tests by adding wildcarded users in the same authCfg then cheking that a user that match the wildcard will have a {expResponse:   okResponse, expStatusCode: http.StatusOK} and a user that doesn't match the wildcard will have a {			expResponse:   "invalid username or password for user \"fooo\"", expStatusCode: http.StatusUnauthorized}
It's not perfect because it doesn't check that the right user/pwd are sent to the fake clickhouse (that is simulated here https://github.com/ContentSquare/chproxy/blob/master/proxy_test.go#L658). You could modify the fake CH so that it returns back in http header the user/pwd sent to him then you can check them in your test but I don't know if it's worth the extra complexity on this already complex file (it's up to you).

**Comment by ilejn** [comment]
Thanks for suggestions.
The implementation is as primitive as possible, just split on '_'.
Reason - I've hesitated to bring trie (AKA prefix tree) to the project, ordered map is not common for golang and some performance degradations could be theoretically noticebale if it is used.

File: config/config.go
Comment (mga-chka) [inline config/config.go]: Regarding hearbeats we recently did a change to use the /ping endpoint of clickhouse so that, by default (if the user doesn't change the request string), chproxy doesn't need a user/pwd anymore to send hearbeats to clickhouse https://github.com/ContentSquare/chproxy/pull/214.

so you should change your conditions (to check is it's the '/ping or not) & modify all the comments saying that by default the first user in clusters.users will be used

File: docs/content/en/use-cases/wildcarded.md
Comment (mga-chka) [inline docs/content/en/use-cases/wildcarded.md:19]: in order to add  extra safety (because people could you crazy login involving "_*" even if it's very unlikely) and to make the feature not rely on naming convention, you should add an optional boolean field is_wildcarded (false by default) to activate or not the feature for the user.
Doing so would also make the function  getUser() a bit faster in the worst cases (many users and a few wildcarded) since you could skip useless string manipulation operations.

**Comment by mga-chka** [comment]
regarding your explanation of the use of the split and why this feature is 'so' limited, I agree with you. 
You're currently the only one using it and it answer your needs. If other people ask for an improvement then we will start thinking about adding more complexity or using an inefficient yet simple way to answer their needs.

File: docs/content/en/use-cases/wildcarded.md
Comment (ilejn) [reply docs/content/en/use-cases/wildcarded.md:19]: Hello, it's a valid point, but are you sure that we need an explicit flag? From performance standpoint it is enough to implicitely switch this feature if a wildcarded user is actually present in config.
If the answer is "yes, we do need an explicit flag", I think we should raise an error if the flag is not set and a wildcarded user is found in config, right?

**Comment by ilejn** [comment]
>  You could modify the fake CH so that it returns back in http header the user/pwd sent to him then you can check them in your test but I don't know if it's worth the extra complexity on this already complex file (it's up to you).

Thanks, very helpfull.
I think it is crucial to have a test that guarantees that wildcarded user machinery works, so yes, I'll try to introduce some feedback in  fake CH.

**Comment by mga-chka** [comment]
Hi,
Thanks for the new commit.
Regarding the use of HasWildcarded, I was more thinking of having a boolean field:
- available from the yml conf
- for each user
-  this field should be at true in order to apply the wildcard logic for this user
-  (the name could be is_wildcarded)

It's a bit of over-engineering but the rational is that if (or more "when" since some people upvoted your issue) new people use the transparent user feature, they might want better flexibility than having usernames that follow the pattern prefix_[wildcard]. For example they might want prefix[wildcard] or [wildcard]suffix or [wildcard]token[wildcard]. Having this is_wildcarded flag would be more convenient because we could not rely on naming convention of users like what you do in your PR. Moreover, in this case the * character would just be interpreted as a wildcard.
So my fear is that if we don't put this is_wildcarded per user now, we might need it within a few months and it would be a backward incompatible change (since previous users of the feature like you would need to change their configuration to make the feature work again).

**Comment by ilejn** [comment]
Hello Christophe, to be honest, I haven't got the idea.

- If a particular user has a wildcard in the name, the flag 'is_wildcarded' seems redundant.
- If a particular user does not have a wildcard in the name, the flag 'is_wildcarded' seems useless.

This flag can be useful
- if we care about the symbol asterisk (and in theory some other symbols) and want to have a user 'notwildcarded_*' with this exact name without matching
- if we assume that only postfix can be substituted, we can specify the left part only (if user 'analyst' is marked as 'is_wildcarded', then 'analyst_jane' is valid.

Could you elaborate, please?

**Comment by mga-chka** [comment]
Sorry, I was not clear enough.

My point was that adding an explicit boolean flag per user was more future proof since it would allow us to add more complex pattern matching later (if asked by other users) and not just the "prefix_*" pattern that answers your usecase.
Because with this flag, the chproxy admin will be able to say whether the * character must be interpreted as a wildcard or not.

Let's say we merge your PR as it is.
Now let's say that in two months someone wants to activate the transparent user feature for usernames that follows the pattern [firstname].[lastname]-Analyst so basically the usernames that matches \*.\*-Analyst.
We will have to add this flag is_wildcarded in the yml conf 
users:
  - name: "\*.\*-Analyst"
  - is_wildcarded: true
  - ...
 and on chproxy part, this user "\*.\*Analyst" will be considered as a pattern to follow with * being a wildcard

If we do this change and you want later to upgrade chproxy you will face authentication issues since with your current PR your end users will be configured without the is_wildcarded flag (that is now needed in the new version your upgraded to). For example
users:
  - name: "analyst_*"
  - ...
  - name: "support_*"
  - ...

**Comment by ilejn** [comment]
Hello Christophe, actually I don't see why introducing [firstname].[lastname]-Analyst leads to the necessity to have an explicit flag,
while it is not important.
Because I do agree with you that reasons to have an explicit flag are possible (e.g. if care about '*' itself).
So, yes, handling it.

The only (at the moment ;) ) question I have: what if a user name does not contain a wildcard, but is marked as is_wildcarded?
Possible options:

1. raise error
2. use password from original query (as I do for wildcarded users), but require exact match.

I think (1) is better.

**Comment by mga-chka** [comment]
We should raise an error

**Comment by mga-chka** [comment]
nb: I agree that my last comment was not explicit "again" because there is something implicit that I didn't say.
We will have to add an explicit flag for \*.\*-Analyst because with your current PR, you transform implicitly a user to a wildcarded user if its name follows exactly this pattern [some_prefix]_[the * character]. Even if I'm not a big fan of implicit conversions, it's very unlikely that a another company using chproxy has usernames that follow exactly this pattern so I'm ok to merge it.
But, when the new \*.\*-Analyst pattern arrives, we will have to think broader and say that now the * character is a wildcard (because another company will ask for the pattern US-\*-Analyst then another for US-Analyst-* and so on). And if we still rely on an implicitly conversion from a user to a wildcarded user, every company using the * character in their usernames will activate the feature without knowing it. And, even if I would never use * character in a username, I'm sure some companies are doing that for some obscure reasons. So, in order to avoid side-effects, we will have to add a explicit flag to say whether or not this user is wildcarded or not.
I hope it's a bit clearer.

**Review by mga-chka — APPROVED** [review]
Hi, the PR looks good.

One last thing (not blocking, do it if you can), can you add a brief description of the feature in this documentation file chproxy/docs/content/en/configuration/users.md ? It can be short since there is already something in wildcarded.md.
My bad I forgot to ask it because I didn't remember it was present -_-.

Since it's an important PR, I'll wait 1-2 days before merging it so that another maintainer can see it if he wants.

**Comment by ilejn** [comment]
I've added a couple of sentences. Looks like my writing skills are not with me Today ;(

File: config/config.go
Comment (gontarzpawel) [inline config/config.go]: Why is it commented?

Comment (ilejn) [reply config/config.go]: It is commented out because we do not need this flag for cluster user at config level, it is in clusterUser struct in scope.go
Removed. 
Thank you for noticing.

**Comment by gontarzpawel** [comment]
Hi @ilejn , one question I've got is why do we need to introduce a new user for heartbeat? I don't understand the need for this change

File: config/config.go
Comment (gontarzpawel) [inline config/config.go]: can you also move it before XXX? 
```suggestion
	HasWildcarded bool `yaml:"has_wildcarded,omitempty"`
```

File: config/config.go
Comment (gontarzpawel) [inline config/config.go]: What's the purpose of it?

Comment (ilejn) [reply config/config.go]: > Hi @ilejn , one question I've got is why do we need to introduce a new user for heartbeat? I don't understand the need for this change

Hello @Garnek20 ,
actually it is done per 
https://github.com/ContentSquare/chproxy/pull/170#discussion_r901602012
" I will create a separate issue to add a specific user for heartbeats and fallback to the default one if not provided".

**Comment by mga-chka** [comment]
@Garnek20: My 2 cents regarding the hearbeats & userNeeded in this PR.

By default we don't need a user (since your recent PR).
But if the chproxy admin decides to use something else than the "/ping" for healthcheck, a user is needed (in this case, the UserNeeded flag in config.go will be automatically be set to true, it just some syntactic sugar to make the code easier to read) . Before this PR, the user selected for healthcheck was automatically the first user of the conf. But with the transparent/wildcarded user feature, this first user might be a transparent user (so only a pattern, not a real user) and the healthcheck won't be able to use it. We could say that in this case chproxy could check for the next user but in the worst case scenario we could have only transparent users in the conf. So this PR adds this optional user for heartbeat in order to work even if the first user is a transparent one.

File: config/config.go
Comment (ilejn) [reply config/config.go]: > What's the purpose of it?

We use basic auth for heartbeat requests if the query is not '/ping'.
I thought that it is better to determine once if user credentials must be included in heartbeat request than to have a condition  'if hb.request != "/ping"' spreaded in the code.
Should I get rid of this flag?

**Comment by ilejn** [comment]
Hm ... I've occasionally commited the suggested change
```
-	HasWildcarded bool
+	HasWildcarded bool `yaml:"has_wildcarded,omitempty"`
```

Honestly I doubt that it is good idea.
If we have no users with is_wildcarded, but set HasWildcarded to true, it does not make sense

File: config/config.go
Comment (ilejn) [reply config/config.go]: > can you also move it before XXX?

Should I move networkReg as well ?

**Comment by mga-chka** [comment]
> Hm ... I've occasionally commited the suggested change
> 
> ```
> -	HasWildcarded bool
> +	HasWildcarded bool `yaml:"has_wildcarded,omitempty"`
> ```
> 
> Honestly I doubt that it is good idea. If we have no users with is_wildcarded, but set HasWildcarded to true, it does not make sense

I agree. the HasWildcarded is meant to be automatically set while loading the conf. Many its place is not in config.go (same for UserNeeded) but I don't see where you be its place (without creating to much over-engineering)

**Comment by gontarzpawel** [comment]
> > Hm ... I've occasionally commited the suggested change
> > ```
> > -	HasWildcarded bool
> > +	HasWildcarded bool `yaml:"has_wildcarded,omitempty"`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Honestly I doubt that it is good idea. If we have no users with is_wildcarded, but set HasWildcarded to true, it does not make sense
> 
> I agree. the HasWildcarded is meant to be automatically set while loading the conf. Many its place is not in config.go (same for UserNeeded) but I don't see where you be its place (without creating to much over-engineering)

Why is it then filled in the configuration here https://github.com/ContentSquare/chproxy/pull/219/files#diff-373f5578b8f362a364b50b0d01919a57da64c128547ca10d49df8ae422237c82R861? I didn't understand it well before, I'm not in favour of that approach, because `HasWildcarded` can be specified by configuration file but as I understand it shouldn't.. Can't we evaluate it elsewhere than configuration?

**Comment by ilejn** [comment]
Hello Pawel, 

I had prepared the whole bunch of objections, ;)
but it turned out that you are right and it is pretty easy to get rid of HasWildcarded in config.go

**Comment by gontarzpawel** [comment]
> Hello Pawel,
> 
> I had prepared the whole bunch of objections, ;) but it turned out that you are right and it is pretty easy to get rid of HasWildcarded in config.go

Hello Ilya,
That's nice, thank you. I feel like UserNeeded serves similar purpose (automatically set). Can we apply similar strategy? 
After that I think that I'll be good with the change :)

**Comment by ilejn** [comment]
Hello Pawel,

>I feel like UserNeeded serves similar purpose (automatically set).

Exactly.

> Can we apply similar strategy? 

Yes, but at a price of coming back to explict request == "/ping" check. I've failed to find a way to keep this flag anywhere except config.

Done.

**Review by gontarzpawel — APPROVED** [review]
Looks good to me!

**Comment by gontarzpawel** [comment]
Let's just fix the linter @ilejn. After that I'll merge it

**Comment by gontarzpawel** [comment]
Thank you @ilejn for the contribution! :)

**Comment by ilejn** [comment]
Hello @Garnek20 , @mga-chka 
thank you for chproxy and for accepting this PR!

BTW, as far as I remember, Contentsquare has a product that optimizes ClickHouse queries.
Am I mistaken?

**Comment by mga-chka** [comment]
more or less.
In contentsquare we created our own DSL (in scala) to automatically transforms business concepts into a relational DSL then we transform them into plain SQL queries. During  transformation form relational DSL into plain SQL queries we're applying some optimization rules. The most important one being the ability to merge multiple subqueries into a single one. 
For example instead of doing
`where id in (select id from tableX where cond1) or id in (select id from tableX where cond1)` 
the optimizer will do
`where id in (select id from tableX where cond1 or cond2)`
there are other rules to merge subqueries and it improves our worst case scenarios by a factor 5-10 (I don't remember exactly the numbers because I did the first version in 2018).
If everything goes well, we will open source the part that goes from a "relationnal DSL" to an optimized plain old query. But be-careful because the rules I did at the time we're mainly focused on our use case (I don't think some people added news rules since) and it's only for scala applications.

FYI, we'll also open source a new product in a few months (if everything goes well...) that is the schema creation & update in clickhouse. Let's say you want to add/delete a column in product, this tools will allow you to test it in another env then do it in prod without manual actions. It was a big pain in ContentSquare because those manual actions led to some mistakes in production.

**Comment by ilejn** [comment]
@mga-chka , thanks.
Yes, I sort of remember about this scala thing.

I am at the point of starting developing a, say, retrospective optimizer. Cost based retrospecive optimizer ;) 
The idea is to analyze query_log to find out problematic queries. My favorite example - queries with JOINs with suboptimal order of tables. I do not dare to compete with Volcano Planner, but some simplest cases can be covered.

Would be great to start from something, not from nothing.

**Comment by DenisGee** [comment]
so we dont have a suffix or prefix just login like "Frodo" in our ldap, is it possible to use this functionality without suffix or prefix just "*" as a login? =)

**Comment by mga-chka** [comment]
for the moment not. 
But as discussed with Ilejn, the design is loose enough to  easiliy add the "*" usecase.
Since we need it at contentsquare, I'll do the implementation (can't say when, if you're un an hurry you can do it). 
I just create an issue so that you can keep track of the feature:
https://github.com/ContentSquare/chproxy/issues/249

**Pull request merged**
